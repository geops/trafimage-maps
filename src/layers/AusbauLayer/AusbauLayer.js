import qs from "query-string";
import MapboxStyleLayer from "../MapboxStyleLayer";

/**
 * Layer for ausbau
 * Extends {@link https://mobility-toolbox-js.geops.io/doc/class/build/ol/layers/MapboxStyleLayer%20js~MapboxStyleLayer%20html-offset-anchor}
 * @private
 * @class
 * @param {Object} [options] Layer options.
 */
class AusbauLayer extends MapboxStyleLayer {
  static updateUrl(filterValue) {
    const params = qs.parse(window.location.search);

    if (filterValue) {
      params.showFilter = filterValue;
      window.history.replaceState(
        undefined,
        undefined,
        `${window.location.pathname}?${qs.stringify(params)}`,
      );
    } else {
      params.showFilter = "true";
    }
  }

  constructor(options = {}) {
    super(options);

    this.initialFiltersById = {};

    this.showFilterParam = qs.parse(window.location.search).showFilter;

    // List of filters availables
    this.filters = [
      { value: "", key: "Projekte im Bau" },
      { value: "2030", key: "Fertigstellung bis 2030" },
      { value: "2035", key: "Fertigstellung bis 2035" },
    ];
    this.filter =
      this.filters.find((filter) => filter.value === this.showFilterParam) ||
      this.filters[0];
  }

  addDynamicFilters() {
    this.applyNewFilter(this.filter.value);
  }

  applyNewFilter(filterValue) {
    const { maplibreMap } = this.mapboxLayer;
    const style = maplibreMap.getStyle();
    if (!maplibreMap || !style) {
      return;
    }
    this.filter = this.filters.find((filter) => filter.value === filterValue);
    const filterToApply = filterValue !== undefined && [
      "==",
      filterValue,
      ["get", "angebotsschritt"],
    ];

    style.layers.forEach(({ id }) => {
      // This filter must affect only Ausbau layers.
      if (!/ausbau/.test(id)) {
        return;
      }

      // Store the initial filter values for each style layer.
      if (!this.initialFiltersById[id]) {
        this.initialFiltersById[id] = maplibreMap.getFilter(id) || [];
      }

      let newFilter = this.initialFiltersById[id].length
        ? [...this.initialFiltersById[id]]
        : null;

      if (!newFilter && filterToApply) {
        newFilter = ["all", filterToApply];
      } else if (filterToApply && newFilter[0] === "all") {
        // we assume the filter is defined with 'all' keyword
        // ['all', filter1, filter2]
        newFilter.push(filterToApply);
      }
      maplibreMap.setFilter(id, newFilter);
    });

    // Update showFilter parameter
    AusbauLayer.updateUrl(filterValue);
  }
}

export default AusbauLayer;
