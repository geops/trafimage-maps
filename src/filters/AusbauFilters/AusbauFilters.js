import React, { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import {
  Select,
  MenuItem,
  FormControl,
  InputLabel,
  makeStyles,
} from '@material-ui/core';
import AusbauLayer from '../../layers/AusbauLayer';
import { updateConstructions } from '../../config/layers';

const useStyles = makeStyles(() => ({
  formControl: {
    marginTop: 10,
    marginBottom: 20,
  },

  selectIcon: {
    marginRight: -4,
  },
}));

const propTypes = {
  layer: PropTypes.instanceOf(AusbauLayer),
};

const defaultProps = {
  layer: null,
};

const FILTER_PROP = 'angebotsschritt';

const AusbauFilters = ({ layer }) => {
  const classes = useStyles();
  const { t } = useTranslation();
  const [value, setValue] = useState('');

  const onChange = useCallback(
    (evt) => {
      const newValue = evt.target.value;
      setValue(newValue);
      layer.applyNewFilter(['==', newValue, ['get', FILTER_PROP]]);
      updateConstructions(layer.mapboxLayer.mbMap);
    },
    [layer, setValue],
  );

  if (!layer || !layer.isShowFilter) {
    return null;
  }

  return (
    <>
      <FormControl fullWidth className={classes.formControl}>
        <InputLabel shrink>{t(FILTER_PROP)}</InputLabel>
        <Select
          autoWidth
          displayEmpty
          value={value}
          onChange={onChange}
          classes={{ icon: classes.selectIcon }}
        >
          <MenuItem value="">{t('Projekte im Bau')}</MenuItem>
          <MenuItem value="2030">{t('Fertigstellung bis 2030')}</MenuItem>
          <MenuItem value="2035">{t('Fertigstellung bis 2035')}</MenuItem>
        </Select>
      </FormControl>
    </>
  );
};

AusbauFilters.propTypes = propTypes;
AusbauFilters.defaultProps = defaultProps;

export default AusbauFilters;
