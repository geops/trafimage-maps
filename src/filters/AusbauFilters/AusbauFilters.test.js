import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '@mui/material';
import theme from '../../themes/default';
import AusbauLayer from '../../layers/AusbauLayer';
import AusbauFilters from '.';

describe('AusbauFilters', () => {
  let layer;

  beforeEach(() => {
    layer = new AusbauLayer();
  });

  describe('should match snapshot', () => {
    test('if no layer defined, should display null.', () => {
      const component = render(
        <ThemeProvider theme={theme}>
          <AusbauFilters />
        </ThemeProvider>,
      );
      expect(component.container.innerHTML).toMatchSnapshot();
    });

    test('if showFilterParam is undefined or null, should display null.', () => {
      const component = render(
        <ThemeProvider theme={theme}>
          <AusbauFilters layer={layer} />
        </ThemeProvider>,
      );
      expect(component.container.innerHTML).toMatchSnapshot();
    });

    test('if showFilterParam is true, should display a select box.', () => {
      layer.showFilterParam = 'true';
      const component = render(
        <ThemeProvider theme={theme}>
          <AusbauFilters layer={layer} />
        </ThemeProvider>,
      );
      expect(component.container.innerHTML).toMatchSnapshot();
    });

    test('if showFilterParam is an empty string, should display a select box.', () => {
      layer.showFilterParam = '';
      const component = render(
        <ThemeProvider theme={theme}>
          <AusbauFilters layer={layer} />
        </ThemeProvider>,
      );
      expect(component.container.innerHTML).toMatchSnapshot();
    });
  });
});
