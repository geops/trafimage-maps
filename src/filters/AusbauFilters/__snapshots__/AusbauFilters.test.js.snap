// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AusbauFilters should match snapshot if no layer defined, should display null. 1`] = `""`;

exports[`AusbauFilters should match snapshot if showFilterParam is an empty string, should display a select box. 1`] = `
<div class="MuiFormControl-root MuiFormControl-fullWidth makeStyles-formControl-4 css-q8hpuo-MuiFormControl-root">
  <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1gis3ws-MuiFormLabel-root-MuiInputLabel-root"
         data-shrink="true"
  >
    angebotsschritt
  </label>
  <div class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1mycgdj-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root">
    <div tabindex="0"
         role="combobox"
         aria-controls=":r1:"
         aria-expanded="false"
         aria-haspopup="listbox"
         class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-1ybbvap-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
    >
      Projekte im Bau
    </div>
    <input aria-invalid="false"
           aria-hidden="true"
           tabindex="-1"
           class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
           value
    >
    <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-1dl46ju-MuiSvgIcon-root-MuiSelect-icon"
         focusable="false"
         aria-hidden="true"
         viewbox="0 0 24 24"
         data-testid="ArrowDropDownIcon"
    >
      <path d="M7 10l5 5 5-5z">
      </path>
    </svg>
    <fieldset aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1lps3wb-MuiOutlinedInput-notchedOutline"
    >
      <legend class="css-ihdtdm">
        <span class="notranslate">
          ​
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`AusbauFilters should match snapshot if showFilterParam is true, should display a select box. 1`] = `
<div class="MuiFormControl-root MuiFormControl-fullWidth makeStyles-formControl-3 css-q8hpuo-MuiFormControl-root">
  <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1gis3ws-MuiFormLabel-root-MuiInputLabel-root"
         data-shrink="true"
  >
    angebotsschritt
  </label>
  <div class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1mycgdj-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root">
    <div tabindex="0"
         role="combobox"
         aria-controls=":r0:"
         aria-expanded="false"
         aria-haspopup="listbox"
         class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-1ybbvap-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
    >
      Projekte im Bau
    </div>
    <input aria-invalid="false"
           aria-hidden="true"
           tabindex="-1"
           class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
           value
    >
    <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-1dl46ju-MuiSvgIcon-root-MuiSelect-icon"
         focusable="false"
         aria-hidden="true"
         viewbox="0 0 24 24"
         data-testid="ArrowDropDownIcon"
    >
      <path d="M7 10l5 5 5-5z">
      </path>
    </svg>
    <fieldset aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1lps3wb-MuiOutlinedInput-notchedOutline"
    >
      <legend class="css-ihdtdm">
        <span class="notranslate">
          ​
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`AusbauFilters should match snapshot if showFilterParam is undefined or null, should display null. 1`] = `""`;
